File: backend/src/application/use-cases/shopping-session/GetShoppingSessionsUseCase.ts
import type { GetShoppingSessionsUseCasePort } from "@application/ports/incoming/GetShoppingSessionsUseCasePort";
import type { ShoppingSessionRepository } from "@application/ports/outgoing/ShoppingSessionRepository";
import type { LoggerPort } from "@application/ports/outgoing/LoggerPort";
import type {
	GetShoppingSessionsInput,
	GetShoppingSessionsResult,
	ShoppingSessionHistoryItem,
} from "./contracts";
import type { ShoppingSessionItem } from "@domain/entities/ShoppingSession";

export class GetShoppingSessionsUseCase
	implements GetShoppingSessionsUseCasePort
{
	constructor(
		private readonly shoppingSessionRepository: ShoppingSessionRepository,
		private readonly logger: LoggerPort,
	) {}

	async execute(
		input: GetShoppingSessionsInput,
	): Promise<GetShoppingSessionsResult> {
		this.logger.info(
			`[GetShoppingSessionsUseCase] execute - userId: "${input.userId}"`,
		);

		const sessions = await this.shoppingSessionRepository.findByUserId(
			input.userId,
		);

		return {
			sessions: sessions.map((session) => ({
				sessionId: session.id,
				shoppedAt: session.shoppedAt.toISOString(),
				totalPrice: session.totalPrice,
				createdAt: session.createdAt.toISOString(),
				items: session.items.map((item) => this.mapSessionItemToOutput(item)),
			})),
		};
	}

	private mapSessionItemToOutput(item: ShoppingSessionItem): ShoppingSessionHistoryItem {
		return {
			productName: item.productName,
			supermarket: item.supermarket,
			category: item.category,
			price: item.price,
			pricePerUnit: item.pricePerUnit,
			unit: item.unit,
			taxType: item.taxType,
			quantity: item.quantity,
			image: item.image,
			url: item.url,
		};
	}
}

File: backend-dotnet/src/api/Application/UseCases/ShoppingSession/GetShoppingSessionsUseCase.cs
using Microsoft.Extensions.Logging;
using SharpMonads.Core;
using SupermarketPlanner.Api.Application.Ports.Driven;
using SupermarketPlanner.Api.Application.Ports.Driving;
using SupermarketPlanner.Api.Domain;
using SupermarketPlanner.Api.Domain.Entities;

namespace SupermarketPlanner.Api.Application.UseCases.ShoppingSession;

public sealed class GetShoppingSessionsUseCase(
    IShoppingSessionRepository shoppingSessionRepository,
    ILogger<GetShoppingSessionsUseCase> logger) : IGetShoppingSessionsUseCase
{
    public Task<Either<DomainError, GetShoppingSessionsResult>> Invoke(
        GetShoppingSessionsInput input,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "[GetShoppingSessionsUseCase] Invoke - userId: \"{UserId}\"",
            input.UserId);

        return shoppingSessionRepository.FindByUserId(input.UserId, cancellationToken)
            .MapAsync(sessions => new GetShoppingSessionsResult(
                sessions.Select(MapSession).ToList()));
    }

    private static ShoppingSessionHistoryEntry MapSession(ShoppingSessionEntity session) =>
        new(
            SessionId: session.Id,
            ShoppedAt: session.ShoppedAt.ToUniversalTime().ToString("o"),
            TotalPrice: session.TotalPrice,
            CreatedAt: session.CreatedAt.ToUniversalTime().ToString("o"),
            Items: session.Items.Select(MapSessionItemToOutput).ToList());

    private static ShoppingSessionHistoryItem MapSessionItemToOutput(ShoppingSessionItem item) =>
        new(item.ProductName, item.Supermarket, item.Category, item.Price,
            item.PricePerUnit, item.Unit, item.TaxType, item.Quantity, item.Image, item.Url);
}
